### PR TITLE
Leap hand hardware

### PR DESCRIPTION
**Pull request by gavin-hyl**
Add and test LEAP hand hardware interface (motor control and reading/publishing joint encoders) into obelisk.

### CODE REVIEW

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/dxl_motor_helper.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/dxl_motor_helper.py:114]: again, there should be whitespace here - make sure the pre-commit stuff is being run (it's probably one reason why CI is failing)

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py:6]: move this below the imports

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py:21]: i'm gonna put these comments here in a centralized place, but you need docstrings on all public classes and methods of public classes as well as public methods. Also, i'm requesting that you use python typing conventions for all arguments and make sure that python type checking passes to the best of your ability. You can check this locally by calling `pyright` in the repo root, and it will tell you what isn't passing (warnings are fine, errors are not).

there might be some things that are very stubborn in terms of the type checking - feel free to ping me on slack to ask about how to resolve. for the truly unresolvable, we can always annotate and make `pyright` ignore.

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_node.py:37]: for now, i think it's ok to leave the setup stuff here. if we need to expand general hardware setup stuff to a separate node for other systems, we can always handle that in a later PR.

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_reset.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_reset.py:1]: where does this get called? shouldn't this be in the `on_shutdown` callback of the `ObeliskLeapHand`?

File: obelisk_ws/src/obelisk_ros/config/dummy_py.yaml
Comment (alberthli) [inline obelisk_ws/src/obelisk_ros/config/dummy_py.yaml:49]: please revert these changes and make a new yaml file specifically for the leap hand with hardware (and probably make another one for sim)

File: obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/.gitignore
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/.gitignore:1]: delete this file

File: obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/obelisk_leap_hand.py
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/obelisk_leap_hand.py:11]: does this executor need to be multithreaded? i don't actually think it does (please discuss multi vs single threaded executors with zach to get an overview)

File: obelisk_ws/src/python/obelisk_leap_py/package.xml
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_leap_py/package.xml:7]: you wrote this, so you should add your name/contact as maintainer (if there's a way to add multiple, maybe add me as well under you)

File: obelisk_ws/src/python/obelisk_leap_py/setup.py
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_leap_py/setup.py:21]: same here, put yourself down as a maintainer for whatever you write

File: obelisk_ws/src/python/obelisk_sim_py/obelisk_sim_py/obelisk_mujoco_robot.py
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_sim_py/obelisk_sim_py/obelisk_mujoco_robot.py:8]: revert these changes and make a new executable file specific for the LEAP hand

File: pixi.toml
Comment (alberthli) [inline pixi.toml:121]: any reason why this version is upper bounded?

File: models/leap/leap_hand.xml
Comment (alberthli) [inline models/leap/leap_hand.xml:1]: on this new `models` directory: i'm pretty sure as it currently stands that if you wanted to run the leap hand examples from a downstream external project (similar to what zach has set up in the examples repo) it would fail, because the downstream repo has no knowledge of these model assets. we should make sure to test this before merging, and figure out a clean solution. one way might be to locate the assets inside the workspace itself. you would need to make sure that those model assets are made available in the package by editing the `package.xml` file of whatever package you put it in (example: either associated with the `obelisk_leap_py` package or even in a generic obelisk package that might only contain model assets of all the lab robots).

**Review by alberthli — CHANGES_REQUESTED** [review]
High-level comments:
- make sure you are linting before committing/asking for review. register the pre-commit hooks and run the hooks locally on all files (google how to do this). after that, they should always run on what you commit automatically.
- don't change existing example code or dummy code. instead, make new files for the specific feature you're implementing, since we're trying to build up a large library of features and examples.

File: docker/docker-compose.yml
Comment (gavin-hyl) [reply docker/docker-compose.yml:47]: Just tested, we don't need to map the serial port if the user is in the group already.

File: pixi.toml
Comment (gavin-hyl) [reply pixi.toml:121]: no...removed it just now

File: obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_reset.py
Comment (gavin-hyl) [reply obelisk/python/obelisk_py/zoo/robot/hardware/leap_hand/leap_reset.py:1]: I originally used this as a test script, but now I added the functionality to the callback function.

File: obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/obelisk_leap_hand.py
Comment (gavin-hyl) [reply obelisk_ws/src/python/obelisk_leap_py/obelisk_leap_py/obelisk_leap_hand.py:11]: Changed to single threaded.

File: obelisk/python/obelisk_py/zoo/control/example/example_position_setpoint_controller.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/control/example/example_position_setpoint_controller.py:1]: you haven't fully reverted the changes to this file - the diff should be empty when compared to `main`.

File: obelisk/python/obelisk_py/zoo/control/example/leap_pos_setpoint.py
Comment (alberthli) [inline obelisk/python/obelisk_py/zoo/control/example/leap_pos_setpoint.py:1]: rename this file to `leap_pos_setpoint_controller.py` - be verbose and explicit about what node type it is when naming.

File: obelisk_ws/src/obelisk_ros/config/dummy_py.yaml
Comment (alberthli) [inline obelisk_ws/src/obelisk_ros/config/dummy_py.yaml:1]: i won't comment on this for the rest of this review, but again, this file has not been fully reverted. you can tell when it is when it no longer appears in this changed files tab on github. go through my old comments asking for reversions and really make sure you've fully reverted changes.

File: obelisk_ws/src/obelisk_ros/config/leap_py.yaml
Comment (alberthli) [inline obelisk_ws/src/obelisk_ros/config/leap_py.yaml:11]: probably remove this "asdf" key that i accidentally committed for testing purposes earlier

File: obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py:1]: move this to the zoo

File: obelisk_ws/src/python/obelisk_control_py/setup.py
Comment (alberthli) [inline obelisk_ws/src/python/obelisk_control_py/setup.py:28]: once moved to the zoo, take this out of this file.

File: pixi.toml
Comment (alberthli) [inline pixi.toml:121]: probably lower bound it still, what i was getting at in the original review is that upper bounding it seemed like a bad default choice

**Review by alberthli — CHANGES_REQUESTED** [review]
some additional minor comments

File: obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py
Comment (gavin-hyl) [reply obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py:1]: If this is moved to the zoo, how can the config file correctly identify where it is? If my understanding is correct right now it looks under the `obelisk_ws/src/python` for packages, and one of the packages must contain an entry point to the controller.

File: models/leap/leap_hand.xml
Comment (gavin-hyl) [reply models/leap/leap_hand.xml:1]: In the config file, I see that the path specified to the xml assumes a root directory of `models` - this feature was implemented before I added the leap hand xml and meshes into it, so I'm not exactly sure how to point obelisk to another directory to look for models. Perhaps I could open up an issue and resolve it in a later PR?

File: obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py
Comment (alberthli) [reply obelisk_ws/src/python/obelisk_control_py/obelisk_control_py/leap_controller.py:1]: I don't remember what it looked like when I made that comment, but it looks good now.

File: models/leap/leap_hand.xml
Comment (alberthli) [reply models/leap/leap_hand.xml:1]: Ah yes, you're right, e.g., [https://github.com/Caltech-AMBER/obelisk/blob/a37f3ff825965ca14d7224396a71352e17f87a51/obelisk/python/obelisk_py/core/sim/mujoco.py#L98-L106](https://github.com/Caltech-AMBER/obelisk/blob/a37f3ff825965ca14d7224396a71352e17f87a51/obelisk/python/obelisk_py/core/sim/mujoco.py#L98-L106). Good catch. I guess this should mean it's fine as long as we assume the end user clones the `model` directory, since we point it there not by using package references, but instead by setting the `OBELISK_ROOT` environment variable.

**Review by alberthli — APPROVED** [review]
LGTM.

**Pull request merged**
